### PR TITLE
[FEAT] Frogs

### PR DESCRIPTION
**Pull request by tiffanydys**
# Description
Frogs is a community driven feature that is made by the community team (Project Dignity : @tiffanydys, @DonatasDee, @VadimasVP, @Paluras)

Discussion #1217 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
`yarn test`
Screen Shot 2022-08-18 at 4 21 41 AM

`yarn dev`
- to test in localhost please DM me your wallet address so I can whitelist you and send you 300 SFL for test minting
- make sure to update your `.env` file from `.env.sample`
- to everyone who wants to view test dummy data of your minted frogs you can go to `src/features/game/actions/loadFrogs.ts` and uncomment line number 15 and comment out line 16:

Screen Shot 2022-08-18 at 4 32 08 AM

### Event Ended

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/features/goblins/merchant/lib/frogMachine.ts
Comment (j923sneaks) [inline src/features/goblins/merchant/lib/frogMachine.ts:181]: can omit `[]`

File: src/features/goblins/merchant/lib/frogMachine.ts
Comment (j923sneaks) [inline src/features/goblins/merchant/lib/frogMachine.ts:208]: based on `mintFrog()`, `verified` always return `true` so i think there's no need for `cond`. also can omit `[]` here in `onDone`

File: src/lib/blockchain/metamask.ts
Comment (j923sneaks) [inline src/lib/blockchain/metamask.ts:335]: cleanup

**Review by j923sneaks — COMMENTED** [review]
Hi @tiffanydys , Awesome job! Will run on my local but here are some initial comments :)

File: src/features/goblins/merchant/MerchantModal.tsx
Comment (tiffanydys) [reply src/features/goblins/merchant/MerchantModal.tsx:161]: Oh right I forgot to update that LOL thanks

File: .env.sample
Comment (j923sneaks) [inline .env.sample:20]: does not match with the one in testnet.yaml

Comment (tiffanydys) [reply .env.sample:20]: Updated this one, thanks!

File: src/features/goblins/merchant/lib/frogMachine.ts
Comment (tiffanydys) [reply src/features/goblins/merchant/lib/frogMachine.ts:208]: You're right, will refactor this thanks!

File: src/features/goblins/merchant/MerchantModal.tsx
Comment (tiffanydys) [reply src/features/goblins/merchant/MerchantModal.tsx]: Button links added ✌🏻

**Comment by veenhdek** [comment]
Tested just now after getting it to launch (.env issues i should have thought of) went pretty smooth I put a full report in discord 😄 https://discord.com/channels/880987707214544966/905567246452129873/1008064550924464149 🚀

File: src/features/goblins/merchant/MerchantModal.tsx
Comment (adamhannigan) [inline src/features/goblins/merchant/MerchantModal.tsx:15]: We can use absolute paths for `assets`.

E.g. `assets/nfts/frogs/frog_unrevealed.gif`

File: src/lib/blockchain/Token.ts
Comment (adamhannigan) [inline src/lib/blockchain/Token.ts:94]: Could we make this more reusable for other community projects and let the caller pass in the contract

I.e. `isTokenApprovedForContract(address: string, attempts = 0):`

File: src/lib/blockchain/metamask.ts
Comment (adamhannigan) [inline src/lib/blockchain/metamask.ts:71]: Since the core team does not have control over this contract we will need an extra `try / catch` clause.

This will be helpful if a community contract decided to destroy the contract (which could cause a whijte screen in the game 😓 )

I.e

```
try {
 this.frog = new Frog(this.web3 as Web3, this.account as string);
} catch (e) {
console.error("Unable to initialise frog contract", e)
}
```

File: src/features/game/actions/loadFrogs.ts
Comment (adamhannigan) [inline src/features/game/actions/loadFrogs.ts:57]: To follow ERC721 standards, we should fetch this URL directly from the smart contract instead of hard coding the `FROG_URL`.

For instance.

`const url = metamask.getFrog().getTokenUri(id)`

This would return the URL from the on-chain data

**Review by adamhannigan — COMMENTED** [review]
Some work in progress feedback 🤗 

Just waiting for a whitelisting (DMed) you and then I can review the UI.

**Comment by adamhannigan** [comment]
@tiffanydys and developers 👋 

One of the aims of introducing community content is to give ownership and scale features quickly. A big part of this is allowing teams to iterate and work at their pace without too much friction - This means there is a risk of the core app breaking due to a bug.

We need to look at architecture/separation of code from the core domain/repo. This will help both sides scale without impacting/breaking the other sides work. Right now where my head is at is creating a dedicated "Community" area that would have it's own map and dedicated "/community" URL. You would still access this through the Merchant on Goblin Land.

We can manage the initial setup of this area with a unique space, but we will need to start moving the code in this PR into a new area `src/features/community`. The community area can connect directly with a user's wallet without needing to go through the typical core authentication flow (this will actually simplify a lot of the features). In terms of refactoring, for example we could move the FrogContract out of `metamask.ts` and into a new `src/features/community/lib/communityContracts.ts` file.

It is a bit of work that I wish I anticipated from the start. Now that I see the code in action, it feels like the right forward to separate these concerns and let each team move at their own pace without fear of breaking one another's code 😅 

Let me create some assets/diagram today to help explain this flow of a "community" area 😄

**Review by spencerdezartsmith — COMMENTED** [review]
Great job so far! Some of the things that I noticed..

*Functionality*
- Once I approved the SFL I was take back to the "approve" modal. I needed to close this and reopen to be able to mint. It would be better to be taken to the "mint" modal automatically after the successful approve tx.

*UI*
- I think we need to do a better job communicating that this is a community feature and what that means. We have spoken about it and Adam is going to reach out with some of the ideas we had for this around creating a seperate community space. 
- I also this it wasn't very clear that I needed to have SFL in my metamask wallet and not my farm wallet. This is a tough message to convey. I did see it in the warning message with a link but I don't think people will click on that tbh. You all might need to think about how to make this message clearer outside of the warning message.
- I also would like to move away from the `error` red colour in the warning messages. We have started using a more traditional warning colour of `orange` which is less threatening 😅 . 
- The style of some of the modals flip our style upside down 🙃 . The whitelisted ones have the text on top and then the image. We typically have it in reverse so I would like to be consistent with that.
- The button should be full width for all of the modals so including the `Mint a frog` one imo. 
- The padding in the modals should be consistent. There is a lot of padding in the minted modals but then very little in the Blacklisted modal . 
- The close button on the modals appears to be `block` and taking up the whole row which looks a little off. 

We are working on setting up some generic modal components to align these styles better to make it easier for devs to not have to "redesign" with every feature. Until we have that we want to try our best to be some what consistent in the design. 

This is coming along nicely though. Can't wait for the release 🐸

**Comment by adamhannigan** [comment]
Here is the PR that implements a community style area.

Ideally we want all the new code/actions etc. to live inside of the `features/community` directory.

I have also introduced a community machine that loads the balance from the **personal wallet** instead of the farm wallet. This means that the HUD will show the amount that is in the personal wallet, which is more ideal than the Goblin HUD which shows the balance from the farm's wallet

https://github.com/sunflower-land/sunflower-land/pull/1348

File: src/lib/blockchain/Token.ts
Comment (tiffanydys) [reply src/lib/blockchain/Token.ts:94]: Good catch! I'll update this as well. ❤️

**Comment by AzaroQL** [comment]
I would propose to make a route to community place from the farm, otherwise it requires two redirects to get there(farm->goblin->community).
How about this place with an arrow and a merchant, same way as for goblin?

File: src/features/goblins/merchant/MerchantModal.tsx
Comment (tiffanydys) [reply src/features/goblins/merchant/MerchantModal.tsx]: Used absolute paths for this file and other files as well 😄

File: src/features/game/actions/loadFrogs.ts
Comment (tiffanydys) [reply src/features/game/actions/loadFrogs.ts]: Hi adam!

The frog URI (metadata URL) is different from the `FROG_URL` (which is the API) I think I would have named it better as FROG_API, I think it would be saver adding it to the .env file rather than adding it to the contract. What do you think? 😅

**Comment by tiffanydys** [comment]
Hey guys thank you so much for the reviews! ❤️

@adamhannigan Thank you for initiating the community garden 😊, I've successfully translated the files to fit the `/community` directory.

@spencerdezartsmith I agree, the modals do have different paddings 😅 I've followed your comments and I've updated the images of the modals on this PR. Please have a look and tell me if there are any more adjustments. ❤️

@AzaroQL Thanks for the inspiration on this. ❤️  I've applied it and moved the yellow book a bit to the bottom. 👀


**Modal to Community Garden:**

File: src/features/game/actions/loadFrogs.ts
Comment (tiffanydys) [reply src/features/game/actions/loadFrogs.ts]: Done! ✌🏻😄

File: src/features/community/CommunityGardenEntry.tsx
Comment (AzaroQL) [inline src/features/community/CommunityGardenEntry.tsx:18]: There are cases when first merchantAudio.play() wont work?

File: src/features/community/CommunityGardenEntry.tsx
Comment (AzaroQL) [inline src/features/community/CommunityGardenEntry.tsx:1]: Shouldnt CommunityGardenEntry and CommunityGardenModal files be placed in src\features\farming\town\components, in the same place as GoblinVillageEntry and GoblinVillageModal?
Those components are displayed on farming screen, so to prevent confusion its better to place them there.

Maybe even create subfolders communityGarden and goblinVillage.

File: src/components/ui/FrogNft.tsx
Comment (AzaroQL) [inline src/components/ui/FrogNft.tsx:13]: There is no modal to open?

File: src/features/community/CommunityGardenEntry.tsx
Comment (AzaroQL) [reply src/features/community/CommunityGardenEntry.tsx:1]: Or other way around, goblin files should be moved out. @adamhannigan whats your opinion on that?

File: src/features/community/merchant/MerchantModal.tsx
Comment (AzaroQL) [inline src/features/community/merchant/MerchantModal.tsx:66]: This link should be opened in new tab.

File: src/features/community/CommunityGardenEntry.tsx
Comment (AzaroQL) [reply src/features/community/CommunityGardenEntry.tsx:18]: Tested it and its actually line 15 that should be removed. IsPlaying check is here to prevent overlaping sound when you open and close modal too fast.

File: src/features/community/CommunityGardenModal.tsx
Comment (AzaroQL) [inline src/features/community/CommunityGardenModal.tsx:41]: This button doesnt work

File: src/features/community/merchant/Merchant.tsx
Comment (AzaroQL) [inline src/features/community/merchant/Merchant.tsx:15]: Check my other comment regarding overlapping audio.

File: src/features/community/components/BackButton.tsx
Comment (AzaroQL) [inline src/features/community/components/BackButton.tsx:13]: Check src\features\goblins\components\TownEntryModal.tsx its better to use that approach for back button. You dont need to pass farmId to that component and it seems faster to load, but that can be just me.

File: src/features/community/merchant/actions/loadFrogs.ts
Comment (adamhannigan) [inline src/features/community/merchant/actions/loadFrogs.ts:15]: Could we wrap all of this logic in a try/catch. Since this will be used in the main application, we need to support cases where the contracts/APIs could be down.

For example.

```
try {
 let frogIds = await communityContracts.getFrog().getFrogIds();
  ...
} catch {
   return null
}
```

File: src/features/community/merchant/actions/loadFrogs.ts
Comment (adamhannigan) [inline src/features/community/merchant/actions/loadFrogs.ts:54]: Screen Shot 2022-08-29 at 8 48 26 am

Calling the `tokenUri` directly on an ERC721 contract will return the full version of the URL. This means you don't need to do concatenation below.

**Review by adamhannigan — COMMENTED** [review]
Nice work @tiffanydys 👏 

There are a few comments by @AzaroQL that need to be addressed. I have included a suggestion for using the `tokenURI` and also adding a `try/catch` around some of the core logic since it will live inside of the game.

Once these are satisfied, let's move forward and launch the frogs 🚀 

Before adding the mainnet address, could you please verify the source code on PolygonScan - if you visit the contract address you will be able to view the contract and upload the source code 😄

**Comment by adamhannigan** [comment]
Screen Shot 2022-08-29 at 8 56 34 am

Btw, very happy with my frog! ❤️

File: src/components/ui/FrogNft.tsx
Comment (tiffanydys) [reply src/components/ui/FrogNft.tsx:13]: Not yet there isn't 👀 (there will be in my new update)

File: src/features/community/CommunityGardenEntry.tsx
Comment (tiffanydys) [reply src/features/community/CommunityGardenEntry.tsx]: removed extra `merchantAudio.play()` thanks!

File: src/features/community/CommunityGardenEntry.tsx
Comment (tiffanydys) [reply src/features/community/CommunityGardenEntry.tsx]: refactored and moved files to `src/features/farming/town/components` thanks!

File: src/features/community/merchant/MerchantModal.tsx
Comment (tiffanydys) [reply src/features/community/merchant/MerchantModal.tsx]: applied thanks!

File: src/features/community/merchant/Merchant.tsx
Comment (tiffanydys) [reply src/features/community/merchant/Merchant.tsx]: removed thanks!

File: src/features/community/components/BackButton.tsx
Comment (tiffanydys) [reply src/features/community/components/BackButton.tsx]: Thanks for this, I never thought of using navigation for it, much cleaner and yes much faster than passing the prop ✌🏻😄

File: src/features/community/merchant/actions/loadFrogs.ts
Comment (tiffanydys) [reply src/features/community/merchant/actions/loadFrogs.ts]: added a try catch on load frogs, thanks ✌🏻

File: src/features/community/merchant/actions/loadFrogs.ts
Comment (tiffanydys) [reply src/features/community/merchant/actions/loadFrogs.ts]: I forgot there's a tokenURI function 😅 thanks adam

**Comment by AzaroQL** [comment]
@tiffanydys Due to recent changes in main branch you will need to apply same fix for community garden.
https://github.com/sunflower-land/sunflower-land/pull/1392/files

File: .github/workflows/testnet.yml
Comment (adamhannigan) [inline .github/workflows/testnet.yml:43]: Is the testnet address meant to be the same as the mainnet address?

**Review by adamhannigan — APPROVED** [review]
@tiffanydys LGTM. There are a few small tweaks but I feel releasing and getting feedback will be more important for the project at this stage. 

What timeline were you looking at to launch this feature?

File: src/components/ui/FrogNft.tsx
Comment (tiffanydys) [reply src/components/ui/FrogNft.tsx]: On second thought, it doesn't need a modal so I've removed it 😅. Thanks!

File: .github/workflows/testnet.yml
Comment (tiffanydys) [reply .github/workflows/testnet.yml:43]: Yes, they're the same contract:

Testnet address:
https://mumbai.polygonscan.com/address/0x52a5ddff59e0a27a1cba75c8045a434b7d296a6c

Mainnet address:
https://polygonscan.com/address/0x52a5ddff59e0a27a1cba75c8045a434b7d296a6c

I didn't change anything from the code on the testnet, I think that's why it generated the same address 😅

**Comment by tiffanydys** [comment]
> @tiffanydys LGTM. There are a few small tweaks but I feel releasing and getting feedback will be more important for the project at this stage. 
> 
> 
> 
> What timeline were you looking at to launch this feature?

hey @adamhannigan this code is up for review 🤗 though i'm not sure as to why the community garden doesn't scroll to the sides when viewed on mobile, any ideas why it doesn't work? 😅

File: src/features/community/CommunityGardenModal.tsx
Comment (tiffanydys) [reply src/features/community/CommunityGardenModal.tsx]: Hey @AzaroQL can you help me on this? I can't seem to make it work whatever I try lmao

Comment (AzaroQL) [reply src/features/community/CommunityGardenModal.tsx]: You need to move modal out of div that has onClick event.

Comment (tiffanydys) [reply src/features/community/CommunityGardenModal.tsx]: Solved, thanks! 😄❤️

**Review by adamhannigan — APPROVED** [review]
Great work @tiffanydys 👏 

I found the issue with the map scrolling and made a few adjustments. It required a new container and some positioning. I have pushed this changes in

You may want to check the new positioning of the seal + aquatics to ensure they are now in the right spot. Otherwise happy to merge this in.

Do you have an exact timeline for when the feature will go live and when the whitelisting will begin?

Once you share those dates we can merge this in and align on announcements for the community 😄

**Comment by DonatasDee** [comment]
> Great work @tiffanydys 👏
> 
> I found the issue with the map scrolling and made a few adjustments. It required a new container and some positioning. I have pushed this changes in
> 
> You may want to check the new positioning of the seal + aquatics to ensure they are now in the right spot. Otherwise happy to merge this in.
> 
> Do you have an exact timeline for when the feature will go live and when the whitelisting will begin?
> 
> Once you share those dates we can merge this in and align on announcements for the community 😄

Our whitelist is almost filled up and we were waiting to finish it once we knew when the code is ready for merge, we'd be glad to allow the whitelists to mint this Friday and then do public mint the next day/s. If that's an okay date then we can start organizing around that. ❤️

**Pull request merged**
